Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -1161,79 +1162,85 @@ def try_to_parse_directives(self, node):
-            if arg_error:
-                raise PostParseError(
-                    pos, 'The exceptval directive takes 0 or 1 positional arguments and the boolean keyword "check"')
-            return ('exceptval', (args[0] if args else None, check))
-
-        directivetype = Options.directive_types.get(optname)
-        if len(args) == 1 and isinstance(args[0], ExprNodes.NoneNode):
-            return optname, Options.get_directive_defaults()[optname]
-        elif directivetype is bool:
-            if kwds is not None or len(args) != 1 or not isinstance(args[0], ExprNodes.BoolNode):
-                raise PostParseError(pos,
-                    'The %s directive takes one compile-time boolean argument' % optname)
-            return (optname, args[0].value)
-        elif directivetype is int:
-            if kwds is not None or len(args) != 1 or not isinstance(args[0], ExprNodes.IntNode):
-                raise PostParseError(pos,
-                    'The %s directive takes one compile-time integer argument' % optname)
-            return (optname, int(args[0].value))
-        elif directivetype is str:
-            if kwds is not None or len(args) != 1 or not isinstance(
-                    args[0], (ExprNodes.StringNode, ExprNodes.UnicodeNode)):
-                raise PostParseError(pos,
-                    'The %s directive takes one compile-time string argument' % optname)
-            return (optname, str(args[0].value))
-        elif directivetype is type:
-            if kwds is not None or len(args) != 1:
-                raise PostParseError(pos,
-                    'The %s directive takes one type argument' % optname)
-            return (optname, args[0])
-        elif directivetype is dict:
-            if len(args) != 0:
-                raise PostParseError(pos,
-                    'The %s directive takes no prepositional arguments' % optname)
-            return optname, kwds.as_python_dict()
-        elif directivetype is list:
-            if kwds and len(kwds.key_value_pairs) != 0:
-                raise PostParseError(pos,
-                    'The %s directive takes no keyword arguments' % optname)
-            return optname, [ str(arg.value) for arg in args ]
-        elif callable(directivetype):
-            if kwds is not None or len(args) != 1 or not isinstance(
-                    args[0], (ExprNodes.StringNode, ExprNodes.UnicodeNode)):
-                raise PostParseError(pos,
-                    'The %s directive takes one compile-time string argument' % optname)
-            return (optname, directivetype(optname, str(args[0].value)))
-        elif directivetype is Options.DEFER_ANALYSIS_OF_ARGUMENTS:
-            # signal to pass things on without processing
-            return (optname, (args, kwds.as_python_dict() if kwds else {}))
-        else:
-            assert False
+    def try_to_parse_directive(self, optname, args, kwds, pos, directives_allowing_expression):
```

Comment:
I really don't think we need this extra argument. We call `try_to_parse_directives()` in two places, one that allows expressions and one that does not. Why not just add a value check after the call that does not allow them, and otherwise simply accept them? `try_to_parse_directive()` does not have to know about this at all.

Reply:
OK implemented please have a look